curvefs/src/client/block_device_client.h

```diff
@@ -25,80 +25,148 @@
  public:
     BlockDeviceClient() {}
     virtual ~BlockDeviceClient() {}
 
     /**
      * @brief Initailize client
-     *
-     * @return error code
+     * @param[in] options the options for client
+     * @return error code (CURVEFS_ERROR:*)
      */
-    virtual CURVEFS_ERROR Init(const BlockDeviceClientOptions &options) = 0;
+    virtual CURVEFS_ERROR Init(const BlockDeviceClientOptions& options) = 0;
 
     /**
      * @brief Uninitailize client
-     *
-     * @return error code
      */
-    virtual CURVEFS_ERROR UnInit() = 0;
+    virtual void UnInit() = 0;
 
     /**
-     * @brief read
-     *
-     * @param buf read buffer
-     * @param offset read start offset
-     * @param length read length
-     *
-     * @return error code
+     * @brief Open file specify by filename and owner
+     * @param[in] filename
+     * @param[in] owner owner for filename
+     * @return error code (CURVEFS_ERROR:*)
      */
-    virtual CURVEFS_ERROR Read(char* buf, off_t offset, size_t length) = 0;
+    virtual CURVEFS_ERROR Open(const std::string& filename,
+                               const std::string& owner) = 0;
 
+    /**
+     * @brief Close the fd which init by Open()
+     * @return error code (CURVEFS_ERROR:*)
+     */
+    virtual CURVEFS_ERROR Close() = 0;
+
+    /**
+     * @brief Get file status
+     * @param[in] filename
+     * @param[in] owner owner for filename
+     * @param[out] statInfo the struct for file status
+     * @return error code (CURVEFS_ERROR:*)
+     */
+    virtual CURVEFS_ERROR Stat(const std::string& filename,
+                               const std::string& owner,
+                               BlockDeviceStat* statInfo) = 0;
+
+    /**
+     * @brief Read from fd which init by Open()
+     * @param[in] buf read buffer
+     * @param[in] offset read start offset
+     * @param[in] length read length
+     * @return error code (CURVEFS_ERROR:*)
+     * @note the offset and length maybe not aligned
+     */
+    virtual CURVEFS_ERROR Read(char* buf, off_t offset, size_t length) = 0;
 
     /**
-     * @brief write
-     *
-     * @param fd file descriptor
-     * @param buf write buffer
-     * @param offset write start offset
-     * @param length write length
-     *
-     * @return error code
+     * @brief Write to fd which init by Open()
+     * @param[in] buf write buffer
+     * @param[in] offset write start offset
+     * @param[in] length write length
+     * @return error code (CURVEFS_ERROR:*)
+     * @note the offset and length maybe not aligned
      */
     virtual CURVEFS_ERROR Write(
         const char* buf, off_t offset, size_t length) = 0;
 };
 
 class BlockDeviceClientImpl : public BlockDeviceClient {
  public:
-    explicit BlockDeviceClientImpl(std::shared_ptr<FileClient> fileClient) :
-        fileClient_(fileClient) {}
-    virtual ~BlockDeviceClientImpl() {}
+    explicit BlockDeviceClientImpl(std::shared_ptr<FileClient> fileClient)
+        : fileClient_(fileClient) {}
```

Comment:
move `fileClient` to `fileClient_`.

And you forget set other member variables, before entering the function body, it will call its default constructor for each variable except built-in types, e.g., isOpen is not guaranteed to false, and fd is not guaranteed to 0.
So this will make `if (!isOpen_ || fd_ < 0) {` unexpected.

Either you can init member variables in the constructor initializer list or just give a default value in the member variable definition.

Reply:
> move `fileClient` to `fileClient_`.
> 
> And you forget set other member variables, before entering the function body, it will call its default constructor for each variable except built-in types, e.g., isOpen is not guaranteed to false, and fd is not guaranteed to 0.
> So this will make `if (!isOpen_ || fd_ < 0) {` unexpected.
> 
> Either you can init member variables in the constructor initializer list or just give a default value in the member variable definition.

This is indeed not a qualified constructor, i will fix it.

Reply:
using "const std::shared_ptr<FileClient> &fileClient" as parameter  will be better.

Reply:
> using "const std::shared_ptr &fileClient" as parameter will be better.

done